For this task we have to open .xlsx file which has 64284 rows x 7 columns (Date, Time, Order ID, Customer ID, Coupon Code, Country, Total). This data shows us information about orders made on IL Makiage website. Identification of possible issues is needed.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Let's read this excel file in pandas and look at the table

In [3]:
sales_db = pd.read_excel('/Users/andreysavitskiyy/Downloads/Database_home_case_analyst.xlsx')
df = pd.DataFrame(sales_db)
df

Date      Time     Order ID  Customer ID Coupon Code Country  \
0     2020-10-01  11:52:28    199652320      69987.0         NaN      CA   
1     2020-10-01  11:54:52    199652339      56440.0         NaN      US   
2     2020-10-01  11:57:14    199652338      77646.0         NaN      US   
3     2020-10-01  11:59:26    199652344       6041.0         NaN      US   
4     2020-10-01  12:01:58  199625188-1      43125.0         NaN      US   
...          ...       ...          ...          ...         ...     ...   
64279 2020-11-30  23:34:19    199780389      81869.0         NaN      US   
64280 2020-11-30  23:40:13    199780382      64591.0        5AFG      US   
64281 2020-11-30  23:40:32    199780383      27503.0         NaN      US   
64282 2020-11-30  23:43:24    199780385      95286.0         NaN      US   
64283 2020-11-30  23:49:55    199780388      51538.0        BARA      US   

       Total  
0      51.55  
1      50.55  
2      55.07  
3      91.20  
4      59.19  
...      ...  
64279  55.56  
64280   9.66  
64281  90.55  
64282  52.51  
64283  15.17  

[64284 rows x 7 columns]

Let's look how many NULL values in this table and detect each of them

In [4]:
df_null_values = df.isna().sum()
df_null_values

Date               0
Time               0
Order ID           1
Customer ID        1
Coupon Code    58040
Country            0
Total              0
dtype: int64

There are 58040 without Coupon codes but it is normal. Let's drop this column and look which row(-s) has/have no Order ID or/and Customer ID 

In [5]:
dataframe_without_codes = df.drop(['Coupon Code'], axis=1)
dataframe_without_codes

Date      Time     Order ID  Customer ID Country  Total
0     2020-10-01  11:52:28    199652320      69987.0      CA  51.55
1     2020-10-01  11:54:52    199652339      56440.0      US  50.55
2     2020-10-01  11:57:14    199652338      77646.0      US  55.07
3     2020-10-01  11:59:26    199652344       6041.0      US  91.20
4     2020-10-01  12:01:58  199625188-1      43125.0      US  59.19
...          ...       ...          ...          ...     ...    ...
64279 2020-11-30  23:34:19    199780389      81869.0      US  55.56
64280 2020-11-30  23:40:13    199780382      64591.0      US   9.66
64281 2020-11-30  23:40:32    199780383      27503.0      US  90.55
64282 2020-11-30  23:43:24    199780385      95286.0      US  52.51
64283 2020-11-30  23:49:55    199780388      51538.0      US  15.17

[64284 rows x 6 columns]

In [6]:
rows_with_empty_cells = dataframe_without_codes[dataframe_without_codes.isna().any(axis=1)]
rows_with_empty_cells

Date      Time Order ID  Customer ID Country  Total
20322 2020-10-25  12:16:12      NaN          NaN      US  55.07

Let's drop this row and look how many empty rows in the table left (if any?)

In [7]:
updated_df = dataframe_without_codes.drop(index=20322)

In [8]:
df_null_values = updated_df.isna().sum()
df_null_values

Date           0
Time           0
Order ID       0
Customer ID    0
Country        0
Total          0
dtype: int64

Let's check is there any abnormal big or small values in our table. 

In [9]:
orders_greater_than_1000 = updated_df.query('Total > 500')
orders_greater_than_1000

Date      Time   Order ID  Customer ID Country    Total
158   2020-10-01  15:23:53  199652769      22124.0      US   556.91
1115  2020-10-02  17:58:54  199656197      68390.0      US   556.15
2220  2020-10-04  07:52:43  199669146      75764.0      US   622.01
3968  2020-10-06  14:15:46  199666182      14474.0      US   512.95
5567  2020-10-08  12:04:54  199681722      56105.0      US   555.60
6241  2020-10-09  09:59:15  199684280      88004.0      US   565.70
13604 2020-10-17  17:48:35  199892951      64004.0      US   502.56
21925 2020-10-28  09:43:15  199839799      16002.0      US   615.09
40677 2020-11-10  23:34:18  199806802      52084.0      US   505.52
58683 2020-11-23  12:24:55  199750041      26035.0      US   566.99
59264 2020-11-24  08:57:31  199762151      64879.0      US  1170.00
59831 2020-11-24  20:50:26  199764167      16123.0      CA   505.90

These orders and their totals are seemed to be pretty normal. Let's check abnormal small values which are less than zero. 

In [10]:
orders_less_than_0 = updated_df.query('Total < 0')
orders_less_than_0

Date      Time   Order ID  Customer ID Country  Total
30947 2020-11-04  17:02:47  199862386      31810.0      US -52.95

Since our table for orders and has no returns, I think we don't need a row with negative number of Total. Let's drop it. 

In [11]:
updated_df = updated_df.drop(index=30947)
updated_df

Date      Time     Order ID  Customer ID Country  Total
0     2020-10-01  11:52:28    199652320      69987.0      CA  51.55
1     2020-10-01  11:54:52    199652339      56440.0      US  50.55
2     2020-10-01  11:57:14    199652338      77646.0      US  55.07
3     2020-10-01  11:59:26    199652344       6041.0      US  91.20
4     2020-10-01  12:01:58  199625188-1      43125.0      US  59.19
...          ...       ...          ...          ...     ...    ...
64279 2020-11-30  23:34:19    199780389      81869.0      US  55.56
64280 2020-11-30  23:40:13    199780382      64591.0      US   9.66
64281 2020-11-30  23:40:32    199780383      27503.0      US  90.55
64282 2020-11-30  23:43:24    199780385      95286.0      US  52.51
64283 2020-11-30  23:49:55    199780388      51538.0      US  15.17

[64282 rows x 6 columns]

Summary: The original dataset had some minor issues with blank spaces in certain columns. This could have caused errors and problems in the analysis during further manipulations of the dataset. 
The "coupon code" column will probably not be needed in the analysis either, as it has more than 58040 empty values, which is inconvenient. In my opinion, it was reasonable to remove this column. 
There were few outliers in this table: there were no abnormally large Total values (the largest value was 1170); however, there was one negative value (-52), which is not realistic in our situation. This line has also been removed. 

How we can segment the customers in meaningful groups:

1. We can segment them based on date and time they bought our product. This segmentation will show us three different segments: active users (within 30 days), regular customers (between 30 and 90 days) and inactive customers (more than 90 days since their last order)
2. Based on country. This criteria, according to prodided data, will show us two different groups: orders from the US and orders from Canada. 
3. Based on order value. We can segment customers, as I propose, into three categories: Low-value customers (orders below 50 dollars), Average-value customers (between 50 dollars and 100 dollars) and High-value customers (more than 100 dollars). 

Let's finally change Customer ID dtype from float64 to int and convert our DataFrame to .csv to import it to MySQL Workbench to solve Task 2 :)

In [16]:
updated_df['Customer ID'] = updated_df['Customer ID'].astype(int)
updated_df

Date      Time     Order ID  Customer ID Country  Total
0     2020-10-01  11:52:28    199652320        69987      CA  51.55
1     2020-10-01  11:54:52    199652339        56440      US  50.55
2     2020-10-01  11:57:14    199652338        77646      US  55.07
3     2020-10-01  11:59:26    199652344         6041      US  91.20
4     2020-10-01  12:01:58  199625188-1        43125      US  59.19
...          ...       ...          ...          ...     ...    ...
64279 2020-11-30  23:34:19    199780389        81869      US  55.56
64280 2020-11-30  23:40:13    199780382        64591      US   9.66
64281 2020-11-30  23:40:32    199780383        27503      US  90.55
64282 2020-11-30  23:43:24    199780385        95286      US  52.51
64283 2020-11-30  23:49:55    199780388        51538      US  15.17

[64282 rows x 6 columns]

In [14]:
to_csv_file = updated_df.to_csv('sales_il_makiage.csv', index=False)